usr/man/man3/signal.3
.TH signal 3 
.SH NAME
signal \- simplified software signal facilities
.SH SYNTAX
.nf
.B #include <signal.h>
.PP
.B (*signal(sig, func))()
.B void (*func)();
.fi
.SH DESCRIPTION
The
.PN signal
subroutine
is a simplified interface to the more general sigvec(2)
facility.
.PP
A signal
is generated by some abnormal event,
initiated by a user at a terminal (quit, interrupt, stop),
by a program error (bus error, etc.),
by request of another program (kill),
or when a process is stopped because it wishes to access
its control terminal while in the background.  For further
information, see tty(4).
Signals are optionally generated
when a process resumes after being stopped,
when the status of child processes changes,
or when input is ready at the control terminal.
Most signals cause termination of the receiving process if no action
is taken; some signals instead cause the process receiving them
to be stopped, or are simply discarded if the process has not
requested otherwise.
Except for the SIGKILL and SIGSTOP
signals, the
.PN signal
call allows signals either to be ignored
or to cause an interrupt to a specified location.
The following is a list of all signals with
names as in the include file < signal.h >:
.LP
.nf
.ta \w'SIGVTALRM 'u +\w'15*  'u
SIGHUP	1	Hangup
SIGINT	2	Interrupt
SIGQUIT	3*	Quit
SIGILL	4*	Illegal instruction
SIGTRAP	5*	Trace trap
SIGIOT	6*	IOT instruction
SIGEMT	7*	EMT instruction
SIGFPE	8*	Floating point exception
SIGKILL	9	Kill (cannot be caught or ignored)
SIGBUS	10*	Bus error
SIGSEGV	11*	Segmentation violation
SIGSYS	12*	Bad argument to system call
SIGPIPE	13	write on a pipe with no one to read it
SIGALRM	14	Alarm clock
SIGTERM	15	Software termination signal
SIGURG	16\*b	Urgent condition present on socket
SIGSTOP	17\*d	Stop (cannot be caught or ignored)
SIGTSTP	18\*d	Stop signal generated from keyboard
SIGCONT	19\*b	Continue after stop
SIGCHLD	20\*b	Child status has changed
SIGTTIN	21\*d	Background read attempted from control terminal
SIGTTOU	22\*d	Background write attempted to control terminal
SIGIO	23\*b	I/O is possible on a descriptor (see fcntl(2))
SIGXCPU	24	Cpu time limit exceeded (see setrlimit(2))
SIGXFSZ	25	File size limit exceeded (see setrlimit(2))
SIGVTALRM	26	Virtual time alarm (see setitimer(2))
SIGPROF	27	Profiling timer alarm (see setitimer(2))
SIGWINCH	28\*b  Window size change
SIGUSR1	30	User defined signal
SIGUSR2	31	User defined signal
SIGCLD		System V name for SIGCHLD
SIGABRT		X/OPEN name for SIGIOT
.fi
.PP
The starred signals in the list above cause a core image
if not caught or ignored.
.PP
If
.I func
is SIG_DFL, the default action
for signal
.I sig
is reinstated; this default is termination
(with a core image for starred signals)
except for signals marked with \*b or \*d.
Signals marked with \*b are discarded if the action
is SIG_DFL; signals marked
with \*d cause the process to stop.
If
.I func
is SIG_IGN the signal is subsequently ignored
and pending instances of the signal are discarded.
Otherwise, when the signal occurs
further occurrences of the signal are
automatically blocked and
.I func
is called.
.PP
A return from the function unblocks
the handled signal and
continues the process at the point it was interrupted.
Unlike previous signal facilities, the handler \fIfunc\fP
remains installed after a signal has been delivered.
.PP
If a caught signal occurs
during certain system calls, causing
the call to terminate prematurely, the call
is automatically restarted.
In particular this can occur
during a
.I read
or write(2)
on a slow device (such as a terminal; but not a file)
and during a wait(2).
.PP
The value of
.PN signal
is the previous (or initial)
value of
.I func
for the particular signal.
.PP
After a fork(2) or vfork(2) the child inherits
all signals.
The
execve(2)
system call
resets all caught signals to the default action;
ignored signals remain ignored.
.SH "RETURN VALUE
The previous action is returned on a successful call.
Otherwise, \-1 is returned and 
.I errno
is set to indicate the error.
.SH ERRORS
The
.PN signal
subroutine
will fail and no action will take place if one of the
following occur:
.TP 15
[EINVAL]
The
.I sig
is not a valid signal number.
.TP 15
[EINVAL]
An attempt is made to ignore or supply a handler for SIGKILL
or SIGSTOP.
.TP 15
[EINVAL]
An attempt is made to ignore SIGCONT (by default SIGCONT
is ignored).
.SH "NOTES  (VAX-11)"
The handler routine can be declared:
.EX
handler(sig, code, scp)
.EE
Here
.I sig
is the signal number, into which the hardware faults and traps are
mapped as defined below.  Code is a parameter which is either a constant
as given below or, for compatibility mode faults, the code provided by
the hardware.  The 
.I scp
is a pointer to the
.I "struct sigcontext"
used by the system to restore the process context from before
the signal.
Compatibility mode faults are distinguished from the
other SIGILL traps by having PSL_CM set in the psl.
.PP
The following defines the mapping of hardware traps to signals
and codes.  All of these symbols are defined in < signal.h >:
.LP
.ta \w'     Floating/decimal divide by zero   'u +\w'15*  'u +8n
.nf
   Hardware condition	Signal	Code

Arithmetic traps:
   Integer overflow	SIGFPE	FPE_INTOVF_TRAP
   Integer division by zero	SIGFPE	FPE_INTDIV_TRAP
   Floating overflow trap	SIGFPE	FPE_FLTOVF_TRAP
   Floating/decimal division by zero	SIGFPE	FPE_FLTDIV_TRAP
   Floating underflow trap	SIGFPE	FPE_FLTUND_TRAP
   Decimal overflow trap	SIGFPE	FPE_DECOVF_TRAP
   Subscript-range	SIGFPE	FPE_SUBRNG_TRAP
   Floating overflow fault	SIGFPE	FPE_FLTOVF_FAULT
   Floating divide by zero fault	SIGFPE	FPE_FLTDIV_FAULT
   Floating underflow fault	SIGFPE	FPE_FLTUND_FAULT
Length access control	SIGSEGV	faulting virtual addr
Protection violation	SIGBUS	faulting virtual addr
Reserved instruction	SIGILL	ILL_PRIVIN_FAULT
Customer-reserved instr.	SIGEMT
Reserved operand	SIGILL	ILL_RESOP_FAULT
Reserved addressing	SIGILL	ILL_RESAD_FAULT
Trace pending	SIGTRAP
Bpt instruction	SIGTRAP
Compatibility-mode	SIGILL	hardware supplied code
Chme	SIGSEGV
Chms	SIGSEGV
Chmu	SIGSEGV
.fi
.SH "SEE ALSO"
kill(1), kill(2), ptrace(2), sigblock(2), sigpause(2),
sigsetmask(2), sigstack(2), sigvec(2), setjmp(3), tty(4)


lib/libcV/sys/signal.c
#ifndef lint
static	char	*sccsid = "@(#)signal.c	1.1	(ULTRIX)	3/31/85";
#endif lint

/************************************************************************
 *									*
 *			Copyright (c) 1985 by				*
 *		Digital Equipment Corporation, Maynard, MA		*
 *			All rights reserved.				*
 *									*
 *   This software is furnished under a license and may be used and	*
 *   copied  only  in accordance with the terms of such license and	*
 *   with the  inclusion  of  the  above  copyright  notice.   This	*
 *   software  or  any  other copies thereof may not be provided or	*
 *   otherwise made available to any other person.  No title to and	*
 *   ownership of the software is hereby transferred.			*
 *									*
 *   This software is  derived  from  software  received  from  the	*
 *   University    of   California,   Berkeley,   and   from   Bell	*
 *   Laboratories.  Use, duplication, or disclosure is  subject  to	*
 *   restrictions  under  license  agreements  with  University  of	*
 *   California and with AT&T.						*
 *									*
 *   The information in this software is subject to change  without	*
 *   notice  and should not be construed as a commitment by Digital	*
 *   Equipment Corporation.						*
 *									*
 *   Digital assumes no responsibility for the use  or  reliability	*
 *   of its software on equipment which is not supplied by Digital.	*
 *									*
 ************************************************************************/
/************************************************************************
 *			Modification History				*
 *									*
 *	David L Ballenger, 30-Mar-1985					*
 * 0001	Change definitions of signal handlers from void (*()) to	*
 *	int (*()).  							*
 *									*
 ************************************************************************/

/*
	signal -- system call emulation for 4.2BSD (VAX version)

	last edit:	13-Jan-1984	D A Gwyn

	NOTE:  Although this module is VAX-specific, it should be
	possible to adapt it to other fairly clean implementations of
	4.2BSD.  The difficulty lies in avoiding the automatic restart
	of certain system calls when the signal handler returns.  I use
	here a trick first described by Donn Seeley of UCSD Chem. Dept.
*/

#include	<errno.h>
#include	<signal.h>

extern int	_sigvec();
extern long	sigsetmask();

extern		etext;


typedef int	bool;			/* Boolean data type */
#define	false	0
#define	true	1

static int	(*handler[NSIG])() =	/* "current handler" memory */
	{
	BADSIG				/* initially, unknown state */
	};
static bool	inited = false;		/* for initializing above */

typedef struct
	{
	int		(*sv_handler)();/* signal handler */
	long		sv_mask;	/* signal mask to apply */
	bool		sv_onstack;	/* take on signal stack */
	}	sigvec;			/* for _sigvec() */

static int	catchsig();
static int	ret_eintr();


int	(*
signal( sig, func )			/* returns previous handler */
	)()
	register int	sig;		/* signal affected */
	register int	(*func)();	/* new handler */
	{
	register int	(*retval)();	/* previous handler value */
	sigvec		oldsv;		/* previous state */
	sigvec		newsv;		/* state being set */

	if ( func >= (int (*)())&etext
       /* && func != SIG_DFL && func != SIG_IGN */
	   )	{
		errno = EFAULT;
		return BADSIG;		/* error */
		}

	/* cancel pending signals */
	newsv.sv_handler = SIG_IGN;
	newsv.sv_mask = 0L;
	newsv.sv_onstack = false;
	if ( _sigvec( sig, &newsv, &oldsv ) != 0 )
		return BADSIG;		/* error */

	/* C language provides no good way to initialize handler[] */
	if ( !inited )			/* once only */
		{
		register int	i;

		for ( i = 1; i < NSIG; ++i )
			handler[i] = BADSIG;	/* initialize */

		inited = true;
		}

	/* the first time for this sig, get state from the system */
	if ( (retval = handler[sig - 1]) == BADSIG )
		retval = oldsv.sv_handler;

	handler[sig - 1] = func;	/* keep track of state */

	if ( func == SIG_DFL )
		newsv.sv_handler = SIG_DFL;
	else if ( func != SIG_IGN )
		newsv.sv_handler = catchsig;	/* actual sig catcher */

	if ( func != SIG_IGN		/* sig already being ignored */
	  && _sigvec( sig, &newsv, (sigvec *)0 ) != 0
	   )
		return BADSIG;		/* error */

	return retval;			/* previous handler */
	}


/* # bytes to skip at the beginning of C ret_eintr() function code: */
#define	OFFSET	2			/* for VAX .word reg_mask */

/* PC will be pointing at a syscall if it is to be restarted: */
typedef unsigned char	opcode;		/* one byte long */
#define	SYSCALL		((opcode)0xBC)	/* VAX CHMK instruction */
#define	IMMEDIATE	((opcode)0x8F)	/* VAX immediate addressing */
/* restartable syscall codes: */
/* syscall codes < 64 are assembled in "literal" mode */
#define	READ		((opcode)3)
#define	WRITE		((opcode)4)
#define	IOCTL		((opcode)54)
/* syscall codes > 63 are assembled in "immediate" mode; there is
   actually another 0 byte after the syscall code too */
#define	WAIT		((opcode)84)
#define	READV		((opcode)120)
#define	WRITEV		((opcode)121)

typedef struct				/* VAX-specific format */
	{
	bool		sc_onstack;	/* sigstack state to restore */
	long		sc_mask;	/* signal mask to restore */
	long		sc_sp;		/* sp to restore */
	opcode		*sc_pc;		/* pc to retore */
	long		sc_ps;		/* psl to restore */
	}	sigcontext;		/* interrupted context */


/*ARGSUSED*/
static int
catchsig( sig, code, scp )		/* signal interceptor */
	register int		sig;	/* signal number */
	long			code;	/* code for SIGILL, SIGFPE */
	register sigcontext	*scp;	/* -> interrupted context */
	{
	register int		(*uhandler)();	/* user handler */
	register opcode		*pc;	/* for snooping instructions */
	sigvec			oldsv;	/* previous state */
	sigvec			newsv;	/* state being set */

	/* at this point, sig is blocked */

	uhandler = handler[sig - 1];

	/* UNIX System V usually wants the state reset to SIG_DFL */
	if ( sig != SIGILL && sig != SIGTRAP /* && sig != SIGPWR */ )
		{
		handler[sig - 1] = newsv.sv_handler = SIG_DFL;
		newsv.sv_mask = 0L;
		newsv.sv_onstack = false;
		(void)_sigvec( sig, &newsv, &oldsv );
		}

	(void)sigsetmask( scp->sc_mask );	/* restore old mask */

	/* at this point, sig is not blocked, usually have SIG_DFL;
	   a longjmp may safely be taken by the user signal handler */

	(*uhandler)( sig );		/* user signal handler */

	/* must now avoid restarting certain system calls */
	pc = scp->sc_pc;
	if ( *pc++ == SYSCALL
	  && (*pc == READ || *pc == WRITE || *pc == IOCTL
	   || *pc++ == IMMEDIATE
	   && (*pc == WAIT || *pc == READV || *pc == WRITEV)
	     )
	   )
		scp->sc_pc = (opcode *)((char *)ret_eintr + OFFSET);

	/* return here restores interrupted context */
	}


static int
ret_eintr()				/* substitute for system call */
{
	errno = EINTR;
	return -1;
}
